Recruit-Training/Advanced-Recruit-Training/Telemetry-Teaser/src/app/app.component.ts

```diff
@@ -11,22 +11,39 @@ export class AppComponent {
   public input: any;
   title = 'Telemetry Recruit Training';
   weather = 50;
-  speed = 0;
-  battery = 0;
-
+  speed = undefined;
+  battery = undefined;
+  speedInput = new FormControl(null, [Validators.required, Validators.min(0), Validators.max(90)]);
+  batteryInput = new FormControl(null, [Validators.required, Validators.min(0), Validators.max(100)]);
+  
+  buttonPressed = false;
+  
+  
   onMouseWeather(event: MouseEvent) {
     this.weather = parseInt((event.target as HTMLInputElement).value);
   }
 
   onKeySpeed(event: KeyboardEvent) {
     this.speed = parseInt((event.target as HTMLInputElement).value);
+    this.buttonPressed=false;
   }
 
   onKeyBattery(event: KeyboardEvent) {
     this.battery = parseInt((event.target as HTMLInputElement).value);
+    this.buttonPressed = false;
   }
 
   calculateRange() {
-
+    let s = this.speed;
+    let b = this.battery;
+    let w = this.weather;
+    this.buttonPressed = true;
+    if (s > 0 && s < 90 && b > 0 && b < 100) {
```

Comment:
This is just an edge case but remember the battery could also be equal to 100% (>= instead of >)